### PR TITLE
CDAP-14630 make metrics minimum interval configurable

### PR DESCRIPTION
**Pull request by yaojiefeng**
JIRA: https://issues.cask.co/browse/CDAP-14630
build: https://builds.cask.co/browse/CDAP-DUT6703-1

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java [hunk 1]
```diff
@@ -227,24 +227,23 @@
...
-      }
+    ImmutableMap.Builder<Integer, Long> builder = ImmutableMap.<Integer, Long>builder()
+      .put(60, minRetentionSecs)
+      .put(3600, hourRetentionSecs);
+    if (minimumResolution < 60) {
+      builder.put(minimumResolution, cConf.getLong(Constants.Metrics.MINIMUM_RESOLUTION_RETENTION_SECONDS));
+    }
+    this.resolutionTTLMap = builder.build();
+    FactTableSupplier factTableSupplier = (resolution, ignoredRollTime) -> {
+      // roll time will be taken from configuration todo: clean this up
```
Comment (CuriousVini) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java]: Can you add jira number for it?

File: cdap-common/src/main/resources/cdap-default.xml [hunk 2]
```diff
@@ -2122,10 +2122,19 @@
...
+      resolution will be 60 seconds.
+    </description>
+  </property>
+
+  <property>
+    <name>metrics.data.table.retention.resolution.minimum.seconds</name>
     <value>7200</value>
     <description>
-      Retention resolution in seconds of the 1-second resolution table;
+      Retention resolution in seconds of the minimum resolution table;
```
Comment (CuriousVini) [inline cdap-common/src/main/resources/cdap-default.xml:2137]: we should also mention that this retention will not be applied if min resolution is >=60

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 3]
```diff
@@ -112,7 +113,8 @@
                                    @Assisted MetricsContext metricsContext,
                                    @Assisted Integer instanceId) {
     this(cConf, metricDatasetFactory, messagingService,
-         schemaGenerator, readerFactory, metricStore, topicNumbers, metricsContext, 1000, instanceId);
+         schemaGenerator, readerFactory, metricStore, topicNumbers, metricsContext,
+         cConf.getInt(Constants.Metrics.METRICS_MINIMUM_RESOLUTION) * 1000, instanceId);
```
Comment (CuriousVini) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: should use `TimeUnit` to convert to millis

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 4]
```diff
@@ -151,7 +153,8 @@
     this.metricsFromAllTopics = new LinkedBlockingDeque<>(queueSize);
     this.topicProcessMetaMap = new ConcurrentHashMap<>();
     this.persistingFlag = new AtomicBoolean();
-    this.metricsProcessIntervalMillis = metricsProcessIntervalMillis;
+    // the max sleep time will be 1 min
+    this.metricsProcessIntervalMillis = metricsProcessIntervalMillis < 60000 ? metricsProcessIntervalMillis : 60000;
```
Comment (CuriousVini) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: should make 6000 a static constant

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java [hunk 5]
```diff
@@ -68,8 +68,9 @@ public FactTable getOrCreateFactTable(int resolution) {
     TableProperties.Builder props = TableProperties.builder();
     // don't add TTL for MAX_RESOLUTION table. CDAP-1626
     if (resolution != Integer.MAX_VALUE) {
-      int ttl = cConf.getInt(Constants.Metrics.RETENTION_SECONDS + resolution +
-                               Constants.Metrics.RETENTION_SECONDS_SUFFIX);
+      int ttl = resolution < 60 ? cConf.getInt(Constants.Metrics.MINIMUM_RESOLUTION_RETENTION_SECONDS) :
```
Comment (CuriousVini) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:71]: should make 60 seconds a static constant

**Review by CuriousVini — COMMENTED** [review]
few comments

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java (hunk 1 above)
Comment (yaojiefeng) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java]: This is a comment from several years ago, created JIRA https://issues.cask.co/browse/CDAP-14730

**Review by CuriousVini — COMMENTED** [review]
no further comments from me.

File: cdap-common/src/main/resources/cdap-default.xml [hunk 6]
```diff
@@ -2122,11 +2122,21 @@
   </property>
 
   <property>
-    <name>metrics.data.table.retention.resolution.1.seconds</name>
+    <name>metrics.minimum.resolution</name>
```
Comment (sreevatsanraman) [inline cdap-common/src/main/resources/cdap-default.xml]: ```metrics.minimum.resolution.seconds```

File: cdap-common/src/main/resources/cdap-default.xml [hunk 7]
```diff
@@ -2122,11 +2122,21 @@
...
+    <value>1</value>
+    <description>
+      Minimum metrics resolution in seconds, this value should be less than 60 seconds, otherwise the minimum metrics
+      resolution will be 60 seconds.
+    </description>
+  </property>
+
+  <property>
+    <name>metrics.data.table.retention.resolution.minimum.seconds</name>
```
Comment (sreevatsanraman) [inline cdap-common/src/main/resources/cdap-default.xml]: ```metrics.data.table.retention.minimum.resolution.seconds```

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java [hunk 8]
```diff
@@ -634,11 +634,14 @@
     public static final String METRICS_TABLE_PREFIX = "metrics.data.table.prefix";
     public static final String TIME_SERIES_TABLE_ROLL_TIME = "metrics.data.table.ts.rollTime";
 
+    public static final String METRICS_MINIMUM_RESOLUTION = "metrics.minimum.resolution";
```
Comment (sreevatsanraman) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: ```metrics.minimum.resolution.seconds```

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java [hunk 9]
```diff
@@ -634,11 +634,14 @@
...
+    public static final String MINIMUM_RESOLUTION_RETENTION_SECONDS =
+      "metrics.data.table.retention.resolution.minimum.seconds";
```
Comment (sreevatsanraman) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: ```metrics.data.table.retention.minimum.resolution.seconds```

**Comment by sreevatsanraman** [comment]
Minor comments, Fix and merge.

**Review by sreevatsanraman — APPROVED** [review]

**Pull request merged**